xdsl/transforms/print_to_println.py

```diff
@@ -0,0 +1,179 @@
+    raise ValueError(f"Cannot find printf code for {t}")
+
+
+class PrintlnOpToPrintfCall(RewritePattern):
+    collected_global_symbs: dict[str, llvm.GlobalOp]
+
+    def __init__(self):
+        self.collected_global_symbs = dict()
+
+    def _construct_global(self, val: str):
+        """
+        Constructs an llvm.global operation containing the string.
+        """
+        data = val.encode() + b"\x00"
+
+        t_type = builtin.TensorType.from_type_and_list(i8, [len(data)])
+
+        return llvm.GlobalOp.get(
+            llvm.LLVMArrayType.from_size_and_type(len(data), i8),
+            _key_from_str(val),
+            constant=True,
+            linkage="internal",
+            value=builtin.DenseIntOrFPElementsAttr.from_list(t_type, data),
+        )
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: print.PrintLnOp, rewriter: PatternRewriter, /):
+        format_str = ""
+        args: list[SSAValue] = []
+        casts: list[Operation] = []
+        # make sure all arguments are in the format libc expects them to be
+        # e.g. floats must be promoted to double before calling
+        for part in _format_string_spec_from_print_op(op):
+            if isinstance(part, str):
+                format_str += part
+            elif isinstance(part.typ, builtin.IndexType):
+                # index must be cast to fixed bitwidth (I think) before printing
+                casts.append(new_val := arith.IndexCastOp.get(part, builtin.i64))
+                args.append(new_val.result)
+                format_str += "%li"
+            elif part.typ == builtin.f32:
+                # f32 must be promoted to f64 before printing
+                casts.append(new_val := arith.ExtFOp.get(part, builtin.f64))
+                args.append(new_val.result)
+                format_str += "%f"
+            else:
+                args.append(part)
+                format_str += _format_str_for_typ(part.typ)
+
+        globl = self._construct_global(format_str + "\n")
+        self.collected_global_symbs[globl.sym_name.data] = globl
+
+        rewriter.replace_matched_op(
+            casts
+            + [
+                ptr := llvm.AddressOfOp.get(
+                    globl.sym_name, llvm.LLVMPointerType.opaque()
+                ),
+                llvm.CallOp("printf", ptr.result, *args),
+            ]
+        )
+
+
+@dataclass(frozen=True)
+class AddExternalFunctionDecl(RewritePattern):
+    name: str
+    signature: llvm.LLVMFunctionType
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: builtin.ModuleOp, rewriter: PatternRewriter, /):
+        rewriter.insert_op_at_end(llvm.FuncOp(self.name, self.signature), op.body.block)
+
+
+@dataclass(frozen=True)
+class AddGlobalSymbols(RewritePattern):
+    symbols: Sequence[llvm.GlobalOp]
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: builtin.ModuleOp, rewriter: PatternRewriter, /):
+        rewriter.insert_op_at_end(self.symbols, op.body.block)
```

Comment:
Currrently unused "in favour" of jank in lines 168+. Let me know if this is a better way!